server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java

```diff
@@ -1162,6 +1166,83 @@ public void testShardStats() throws IOException {
+                .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, 1)
+                .build();
+        IndexMetaData metaData = IndexMetaData.builder(shardRouting.getIndexName())
+                .settings(settings)
+                .primaryTerm(0, 1)
+                .build();
+
+        // Override one Directory methods to make it fail at our will
+        // We use AtomicReference here to inject failure in the middle of the test not immediately
+        // We use Supplier<IOException> instead of IOException to produce meaningful stacktrace
+        // (remember stack trace is filled when exception is instantiated)
+        AtomicReference<Supplier<IOException>> exceptionToThrow = new AtomicReference<>();
+        AtomicBoolean throwWhenMarkingStoreCorrupted = new AtomicBoolean(false);
+        Directory directory = new FilterDirectory(newFSDirectory(shardPath.resolveIndex())) {
+            //fileLength method is called during storeStats try block
+            //it's not called when store is marked as corrupted
+            @Override
+            public long fileLength(String name) throws IOException {
+                Supplier<IOException> ex = exceptionToThrow.get();
+                if (ex == null) {
+                    return super.fileLength(name);
+                } else {
+                    throw ex.get();
+                }
+            }
+
+            //listAll method is called when marking store as corrupted
+            @Override
+            public String[] listAll() throws IOException {
+                Supplier<IOException> ex = exceptionToThrow.get();
+                if (throwWhenMarkingStoreCorrupted.get() && ex != null) {
+                    throw ex.get();
+                } else {
+                    return super.listAll();
+                }
+            }
+        };
+
+        Store store = createStore(shardId, new IndexSettings(metaData, Settings.EMPTY), directory);
+        IndexShard shard = newShard(shardRouting, shardPath, metaData, store,
+                null, new InternalEngineFactory(), () -> {}, EMPTY_EVENT_LISTENER);
+
+        recoverShardFromStore(shard);
+        boolean corruptIndexException = true;
+
+        if (randomBoolean()) {
+            exceptionToThrow.set(() -> new CorruptIndexException("Test CorruptIndexException", "Test resource"));
+            throwWhenMarkingStoreCorrupted.set(randomBoolean());
+        } else {
+            corruptIndexException = false;
+            exceptionToThrow.set(() -> new IOException("Test IOException"));
+        }
+        ElasticsearchException e = expectThrows(ElasticsearchException.class, shard::storeStats);
+        assertThat(shard.state(), equalTo(IndexShardState.CLOSED));
+        if (corruptIndexException && !throwWhenMarkingStoreCorrupted.get()) {
+            assertTrue(store.isMarkedCorrupted());
+        }
+
+        //TODO Should it be invoked when closing the shard?
+        try {
+            directory.close();
```

Comment:
@bleskes what do you think about explicitly closing the directory here? I've added this line, because w/o it test fails and reports unclosed resources. But is not it a shard that is responsible for directory lifecycle?